pkg/apis/caching/v1alpha1/image_types.go

```diff
@@ -0,0 +1,164 @@
+// components express a desire to have a particular image cached.
+type Image struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec holds the desired state of the Image (from the client).
+	// +optional
+	Spec ImageSpec `json:"spec,omitempty"`
+
+	// Status communicates the observed state of the Image (from the controller).
+	// +optional
+	Status ImageStatus `json:"status,omitempty"`
+}
+
+// Check that Image can be validated and defaulted.
+var _ apis.Validatable = (*Image)(nil)
+var _ apis.Defaultable = (*Image)(nil)
+
+// ImageSpec holds the desired state of the Image (from the client).
+type ImageSpec struct {
+	Image string `json:"image"`
+
+	// +optional
+	ServiceAccountName string `json:"serviceAccountName,omitempty"`
+
+	// +optional
+	ImagePullSecrets *corev1.LocalObjectReference `json:"imagePullSecrets,omitempty"`
+}
+
+// ImageConditionType is used to communicate the status of the reconciliation process.
+type ImageConditionType string
+
+const (
+	// ImageConditionReady is set when the revision is starting to materialize
+	// runtime resources, and becomes true when those resources are ready.
+	ImageConditionReady ImageConditionType = "Ready"
+)
+
+// ImageCondition defines a readiness condition for a Image.
+// See: https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#typical-status-properties
+type ImageCondition struct {
+	Type ImageConditionType `json:"type" description:"type of Image condition"`
+
+	Status corev1.ConditionStatus `json:"status" description:"status of the condition, one of True, False, Unknown"`
+
+	// +optional
+	// We use VolatileTime in place of metav1.Time to exclude this from creating equality.Semantic
+	// differences (all other things held constant).
+	LastTransitionTime apis.VolatileTime `json:"lastTransitionTime,omitempty" description:"last time the condition transit from one status to another"`
+
+	// +optional
+	Reason string `json:"reason,omitempty" description:"one-word CamelCase reason for the condition's last transition"`
+
+	// +optional
+	Message string `json:"message,omitempty" description:"human-readable message indicating details about last transition"`
+}
+
+// ImageStatus communicates the observed state of the Image (from the controller).
+type ImageStatus struct {
+	// Conditions communicates information about ongoing/complete
+	// reconciliation processes that bring the "spec" inline with the observed
+	// state of the world.
+	// +optional
+	Conditions []ImageCondition `json:"conditions,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ImageList is a list of Image resources
+type ImageList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []Image `json:"items"`
+}
+
+// IsReady looks at the conditions and if the Status has a condition
+// ImageConditionReady returns true if ConditionStatus is True
```

Comment:
This is probably just a little too custom to auto-generate or reflect over, I'm guessing.

Reply:
Yeah, but see [here](https://github.com/knative/serving/issues/1824).